File: chapter15/多核/dekker.c
/*
 * 采用pthread实现Dekker算法。
 *
 * 编译命令：
 * gcc -O2 -o dekker dekker.c -lpthread
 *
 * 要查看执行进度的话，执行命令：
 * gcc -DPRINT_PROGRESS -O2 -o dekker dekker.c -lpthread
 */ 

#include <assert.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>

static volatile int flag1 = 0;
static volatile int flag2 = 0;
static volatile int turn  = 1;
static volatile int gSharedCounter = 0;
int gLoopCount;
int gOnePercent;

void dekker1( ) {
        flag1 = 1;
        turn  = 2;
        while((flag2 ==  1) && (turn == 2)) ;
        // 关键区域 - 已经获取了锁
        gSharedCounter++;
        // 允许另一个线程执行
        flag1 = 0;
}

void dekker2(void) {
        flag2 = 1;
        turn = 1;
        while((flag1 ==  1) && (turn == 1)) ;
        // 关键区域 - 已经获取了锁
        gSharedCounter++;        
        // 离开关键区域并允许另一个线程执行
        flag2 = 0;
}

void *task1(void *arg) {
        int i,j;
        printf("开始task1\n");
#ifdef PRINT_PROGRESS
	for(i=0;i<100;i++) {
	  printf("[task1]的进度：%d%%\n",i);
	  for(j=gOnePercent;j>0;j--) {
	    dekker1();
	  }
	}
#else
        for(i=gLoopCount;i>0;i--) {
                dekker1();
        }
#endif

}

void *task2(void *arg) {
        int i,j;
        printf("开始task2\n");
#ifdef PRINT_PROGRESS
	for(i=0;i<100;i++) {
	  printf("[task2]的进度：%d%%\n",i);
	  for(j=gOnePercent;j>0;j--) {
	    dekker2();
	  }
	}
#else
        for(i=gLoopCount;i>0;i--) {
                dekker2();
        }
#endif
}

int
main(int argc, char ** argv)
{
        int            loopCount = 0;
        pthread_t      dekker_thread_1;
        pthread_t      dekker_thread_2;
        void           * returnCode;
        int            result;
        int            expected_sum;

        if(argc != 2) 
        {
                fprintf(stderr, "USAGE: %s <循环次数>\n", argv[0]);
                exit(1);
        }

        loopCount   = atoi(argv[1]);	
        gLoopCount  = loopCount;
	gOnePercent = loopCount/100;
        expected_sum = 2*loopCount;
        
        /* 启动线程 */
        result = pthread_create(&dekker_thread_1, NULL, task1, NULL);
        result = pthread_create(&dekker_thread_2, NULL, task2, NULL);

        /* 等待线程介绍 */
        result = pthread_join(dekker_thread_1,&returnCode);
        result = pthread_join(dekker_thread_2,&returnCode);
        printf("两个线程执行完毕！\n");

        /* 结果判断 */
        if( gSharedCounter != expected_sum ) {
                printf("[-] Dekker不能用, 得到的实际结果 %d 与期望结果 %d 不同。\n", gSharedCounter, expected_sum);
                printf("    由于内存模型的问题，共有 %d 次更新因竞争访问内存资源而遗漏了！\n", (expected_sum-gSharedCounter));
                return 1;
        } else {
                printf("[+] Dekker算法可用.\n");
                return 0;
        }
}

File: chapter15/多核/Dekker.java
public class Dekker {	
    private static volatile int _flag1 = 0;
	private static volatile int _flag2 = 0;
	private static volatile int _turn = 1;
	private static volatile int _sharedCounter = 0;

    private static void dekker1() {
	    _flag1 = 1;
		_turn = 2;
		while ((_flag2 == 1) && (_turn == 2));
		_sharedCounter++;
		_flag1 = 0;
	}

	private static void dekker2() {
	    _flag2 = 1;
		_turn = 1;
		while ((_flag1 == 1) && (_turn == 1));
		_sharedCounter++;
		_flag2 = 0;
	}
	
	public static void main(String[] args) throws Exception {
	    if ( args.length != 1 ) {
		     System.out.println("使用方法: java Dekker <循环次数>");
			 return;
		}

		final int loopCount = Integer.parseInt(args[0]);
		Thread thread1 = new Thread(new Runnable() {
		    public void run() {
				for ( int i = 0; i < loopCount; ++i ) {
					dekker1();
				}
			}
		});
		Thread thread2 = new Thread(new Runnable() {
		    public void run() {
				for ( int i = 0; i < loopCount; ++i ) {
					dekker2();
				}
			}
		});
		thread1.start();
		thread2.start();

		thread1.join();
		thread2.join();

		int expected_sum = 2 * loopCount;
		if ( _sharedCounter != expected_sum ) {
		     System.out.println(
			     String.format("资源竞争问题: 实际结果 $1%d 不等于期望结果 $2%d", _sharedCounter, expected_sum));
		} else { 
                     System.out.println("Dekker算法可以使用。");
                }
	}
}
